Source/WebCore/editing/FrameSelection.cpp

```diff
@@ -407,8 +407,21 @@ bool FrameSelection::setSelectionWithoutUpdatingAppearance(const VisibleSelectio
     }
 
     // Selection offsets should increase when LF is inserted before the caret in InsertLineBreakCommand. See <https://webkit.org/b/56061>.
-    if (RefPtr textControl = enclosingTextFormControl(newSelection.start()))
-        textControl->selectionChanged(options.contains(SetSelectionOption::FireSelectEvent));
+    // https://www.w3.org/TR/selection-api/#selectionchange-event
+    if (RefPtr textControl = enclosingTextFormControl(newSelection.start())) {
+        if (textControl->selectionChanged(options.contains(SetSelectionOption::FireSelectEvent))) {
+            document->eventLoop().queueTask(TaskSource::UserInteraction, [ textControl = GCReachableRef { *textControl }]() {
+                textControl->dispatchEvent(Event::create(eventNames().selectionchangeEvent, Event::CanBubble::Yes, Event::IsCancelable::No));
+            });
+        }
+    } else {
+        document->eventLoop().queueTask(TaskSource::UserInteraction, [weakDocument = WeakPtr { document.get() }]() {
```

Comment:
Do we want to fire selectionchangeEvent event when willMutateSelection is false?

Reply:
Oh I guess not. Will fix.